### ISSUE TITLE
Ability to use `ITransactionalSession` via dependancy injection in an incoming message context

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.TransactionalSession#65]
Currently `ITransactionalSession` is not available in an incoming message context.

Unfortunately [UniformSession ](https://github.com/Particular/NServiceBus.UniformSession) cannot be used together with TransactionalSession.

Code that could be invoked in and outside an incoming message context should have the abilily to share a common interface that works in both use cases.

**Comment by danielmarbach** [comment]
We discussed this during our syncs. We still believe the UniformSession design has a distinct purpose which is transitioning from v5 to v6. This is further highlighted in our documentation

> It is encouraged to **embrace the design approach introduced in Version 6.** Migrating to this new design in one step can be difficult though. This package reintroduces an opt-in approach for a uniform session approach that works seamlessly as a message session outside the pipeline and as a pipeline context inside the message handling pipeline. **The message operations provided on the uniform session represent a common denominator but do not support more advanced scenarios like persistence session access, subscribe/unsubscribe operations or access to the incoming message's headers**.

https://docs.particular.net/nservicebus/messaging/uniformsession

TransactionalSession and UniformSession should not be convoluted/married together. 

What we can see though is the benefit of the `ITransactionalSession` becoming the new goto interface for inside and outside the message handling pipeline. Technically, it would be possible already today to use the interface also in components inside the message handling pipeline but it was outside the scope of our work to fully validate all the edge cases surrounding that. Since the transactional session is design to behave exactly as the logic in the incoming message handling pipeline things might just work magically but there might be intricacies with both the core interacting with the storage seam as well as the transactional session interacting with the same. But the common interaction point today is the completable-synchronized storage session which is fully managed by the core and should technically already be in the right "state" as long as the transactional session is only used within the invoke handler context stage.

There is definitely value in further assessing the merits of making the transactional session the new "uniform session" (in the sense of uniform but not in the sense of legacy transitioning between v5 and v6)

**Comment by ramonsmits** [comment]
> What we can see though is the benefit of the `ITransactionalSession` becoming the new goto interface for inside and outside the message handling pipeline. Technically, it would be possible already today to use the interface also in components inside the message handling pipeline but it was outside the scope of our work to fully validate all the edge cases surrounding that.

I like your idea of replacing it. I think most users only need a DI friendly/compatible method that works in both situations.

**Comment by pardahlman** [comment]
I think this is an interesting topic, and I have a question relating to it. Consider the following applications:

* A console application running NServiceBus with message handlers (`IHandleMessages<>`), SQL persistence and outbox
* An ASP.NET Core application with SQL persistance, outbox and transactional session

Both projects takes a dependency on a third class library containing business code that reads/writes to database and publishes messages. I guess it is correct to use `ISqlStorageSession` for database access, as it is available through dependency injection, but what APIs should be used for publishing messages? `IMessageHandlerContext` is only available in a "message handler pipeline", `IMessageSession` is not part of the outbox operation. If I understand the above discussion correctly, `ITransactionalSession` might be a perfect fit for this scenario?

Are there any known issues with using `ITransactionalSession` in `IHandleMessages<>`?

**Comment by danielmarbach** [comment]
@pardahlman yes your understanding is correct, and we think it might even work today but haven't fully tested it yet.

**Comment by pardahlman** [comment]
This is indeed exciting. Looking forward to when this is officially supported 🎉

**Comment by timbussmann** [comment]
bumping this issue. This improvement would be highly beneficial for one of our large code bases to enable teams making use of the transactional session with reasonable amount of effort.

**Comment by jbogard** [comment]
I managed to make this work on a client. First I made an adapter interface:

```csharp
public interface ISessionAdapter : IUniformSession
{
}
```

Then made adapters for both `ITransactionalSession` and `IUniformSession`:

```csharp
public class TransactionalSessionAdapter(ITransactionalSession session) : ISessionAdapter
{
    public Task Send(object message, SendOptions sendOptions, CancellationToken cancellationToken = default)
        => session.Send(message, sendOptions, cancellationToken);

    public Task Send<T>(Action<T> messageConstructor, SendOptions sendOptions, CancellationToken cancellationToken = default)
        => session.Send(messageConstructor, sendOptions, cancellationToken);

    public Task Publish(object message, PublishOptions publishOptions, CancellationToken cancellationToken = default)
        => session.Publish(message, publishOptions, cancellationToken);
    
    public Task Publish<T>(Action<T> messageConstructor, PublishOptions publishOptions,
        CancellationToken cancellationToken = default)
            => session.Publish(messageConstructor, publishOptions, cancellationToken);
}

public class UniformSessionWrapper(IUniformSession session) : ISessionAdapter
{
    public Task Send(object message, SendOptions sendOptions, CancellationToken cancellationToken = default)
        => session.Send(message, sendOptions, cancellationToken);

    public Task Send<T>(Action<T> messageConstructor, SendOptions sendOptions, CancellationToken cancellationToken = default)
        => session.Send(messageConstructor, sendOptions, cancellationToken);

    public Task Publish(object message, PublishOptions publishOptions, CancellationToken cancellationToken = default)
        => session.Publish(message, publishOptions, cancellationToken);
    
    public Task Publish<T>(Action<T> messageConstructor, PublishOptions publishOptions,
        CancellationToken cancellationToken = default)
            => session.Publish(messageConstructor, publishOptions, cancellationToken);
}
```

Common services then depend on `ISessionAdapter` instead of `IUniformSession`.

For places where we'd use uniform session (in a message endpoint), we do:

```csharp
endpointConfiguration.EnableUniformSession();
builder.Services.AddTransient<ISessionAdapter, UniformSessionAdapter>();
```

For send-only endpoints where we're using transactional session:
```csharp
var transactionalSessionOptions = new TransactionalSessionOptions { ProcessorEndpoint = "TransactionalSessionProcessor" };
persistence.EnableTransactionalSession(transactionalSessionOptions);

builder.Services.AddTransient<ISessionAdapter, TransactionalSessionAdapter>();
```

I modified the Uniform Session example to validate that these work.
